### ISSUE TITLE
mod_block_strangers: Some possible improvements

### ISSUE THREAD

**Issue by vanitasvitae** [processone/ejabberd#2644]
Environment
-----------
- ejabberd version: 18.09
- Erlang version: `Erlang (SMP,ASYNC_THREADS) (BEAM) emulator version 9.3.3.3`
- OS: Linux (Raspbian)
- Installed from: source

Hi!
I'm using mod_block_strangers to fight some spamming issues on my server.
The module does a great job at blocking spam, however there are some things that might be improved in my opinion.

* The documentation of mod_block_strangers states: `This option specifies if strangers messages should be logged (as info message) in ejabberd logs.` Even though the option `log` is set to true, the actual body of a dropped message is lost. It would be nice to have the body written to the logs as well.
* Strangers are only actively notified about being blocked after they sent a subscription request. I'd like to have strangers being notified about the block as soon as they send a message to me.
* Solving the captcha only unblocks the subscription request, while sending messages is still not possible. It would be nice to have an option to let the captcha also unlock messages.
* The wording of ejabberd_captcha is not very precise. `Your messages to ~s are being blocked. To unblock them, visit ~s` makes it sound as if the captcha unblocks the messages, while in fact it only unblocks the subscription request. This lead to some confused users already. I got around this issue by modifying the text in `src/ejabberd_captcha.erl` though.

**Comment by zinid** [comment]
Regarding the last one: just propose a PR with better wording, so it will also reflect MUC messages for example.
I will address other suggestions a bit later.

**Comment by vanitasvitae** [comment]
Yeah, I'm currently thinking about a good wording.

Regarding MUC messages, I often read, that some messages from MUCs in which I participate are blocked. What might be the cause for this? I assume this is the case, when one of my clients is no longer part of that MUC, but the MUC assumes otherwise. Is my assumption right?

**Comment by zinid** [comment]
> I assume this is the case, when one of my clients is no longer part of that MUC, but the MUC assumes otherwise

Yes, your local server doesn't see you having sent direct presence to that MUC room and thus blocks incoming messages from the MUC room (which in turn sees you as joined).

**Comment by zinid** [comment]
Checking bookmarks is a hacky but working solution to this, BTW. Given that the module itself already is a hack and knows too much about communications, I think we can implement the check.

**Comment by mdosch** [comment]
I got bitten by this as well. When reporting spammers to the 0157 contact I got spammed by captcha requests although I already solved them several times and I had to block that contact to get rid of getting a captcha request every few minutes. I also thought solving the captcha enables me.to write messages and I sent some logs of spam I received. Maybe this accumulated captcha requests which were sent to me delayed (I can't explain otherwise why I still got spammedvl with captchas after I stopped writing to that person).

Also it's pretty bad to apply this to the 0157 abuse contact as I had to block that server as I had no possibility to reach the admin, but that should probably go into another issue.

**Comment by licaon-kter** [comment]
Another data point, user said: _"I didn't add you back immediately, that's why you've been "spammed" by captcha"_
